rudder-core/src/main/scala/com/normation/rudder/api/DataStructures.scala

```diff
@@ -0,0 +1,224 @@
+ */
+final case class ApiToken(value: String) extends HashcodeCaching
+
+object ApiToken {
+
+  val tokenRegex = """[0-9a-zA-Z]{12,128}""".r
+
+  def buildCheckValue(value: String) : Option[ApiToken] = value.trim match {
+    case tokenRegex(v) => Some(ApiToken(v))
+    case _ => None
+  }
+}
+
+
+/*
+ * HTTP verbs
+ */
+sealed trait HttpAction { def name: String }
+final case object HttpAction {
+
+  final case object HEAD   extends HttpAction { val name = "head" }
+  final case object GET    extends HttpAction { val name = "get" }
+  // perhaps we should have an "accepted content type"
+  // for update verbs
+  final case object PUT    extends HttpAction { val name = "put" }
+  final case object POST   extends HttpAction { val name = "post" }
+  final case object DELETE extends HttpAction { val name = "delete" }
+
+  //no PATCH for now
+
+  def values = ca.mrvisser.sealerate.values[HttpAction]
+
+  def parse(action: String): Either[String, HttpAction] = {
+    val lower = action.toLowerCase()
+    values.find( _.name == lower ).toRight(s"Action '${action}' is not recognized as a supported HTTP action")
+  }
+}
+
+/*
+ * An authorization controle is done in a path (so we are managing
+ * tree by essense).
+ *
+ * The path can be composed of 3 kinds of segments (as called in rfc3986):
+ * - a named segment (ie the name of resources)
+ * - a single-segment wildcard (in a glob metaphor, a '*')
+ * - a multi-segment wildcard ('**'). This one can only appear one time
+ *   as the last segment of the path.
+ */
+
+sealed trait AclPathSegment { def value: String }
+
+final object AclPathSegment {
+  final case class  Segment(value: String) extends AclPathSegment
+  final case object Wildcard               extends AclPathSegment { val value = "*" }
+  final case object DoubleWildcard         extends AclPathSegment { val value = "**" }
+
+  def parse(s: String): Either[String, AclPathSegment] = {
+    s.trim() match {
+      case ""   => Left("An ACL path segment can not be empty")
+      case "*"  => Right(Wildcard)
+      case "**" => Right(DoubleWildcard)
+      case x    => Right(Segment(x))
+    }
+  }
+}
+
+/*
+ *
+ */
+sealed trait AclPath {
+  def value: String = parts.toList.map( _.value ).mkString("/")
+  def parts: NonEmptyList[AclPathSegment]
+}
+
+final object AclPath {
+
+ // the full path is enumarated. At least one segment must be given ("/" is not possible
+  // in our simpler case)
+  final case class FullPath(segments: NonEmptyList[AclPathSegment]) extends AclPath {
+    def parts = segments
+  }
+  // only the root is given, and the path ends with "**". It can even be onlyl "**"
+  final case class Root(segments: List[AclPathSegment])              extends AclPath {
+    def parts = NonEmptyList.ofInitLast(segments, AclPathSegment.DoubleWildcard)
+  }
+
+
+  // parse a path to an acl path.
+  // we don't accept empty string and ignore empty subpart,
+  // and "**" must be the last segment.
```

Comment:
This is not checked by the method, should we check it ?

Reply:
done